ruma-events/src/room/message.rs

```diff
@@ -34,8 +34,83 @@ pub type MessageEvent = OuterMessageEvent<MessageEventContent>;
 #[derive(Clone, Debug, Serialize, MessageEventContent)]
 #[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
 #[ruma_event(type = "m.room.message")]
-#[serde(untagged)]
-pub enum MessageEventContent {
+pub struct MessageEventContent {
+    /// A key which identifies the type of message being sent.
+    ///
+    /// This also holds the specific content of each message.
+    #[serde(flatten)]
+    pub msgtype: MessageType,
+
+    /// Information about related messages for
+    /// [rich replies](https://matrix.org/docs/spec/client_server/r0.6.1#rich-replies).
+    #[serde(rename = "m.relates_to", skip_serializing_if = "Option::is_none")]
+    pub relates_to: Option<Relation>,
+
+    /// New content of an edited message.
+    ///
+    /// This should only be set if `relates_to` is `Some(Relation::Replacement(_))`.
+    #[cfg(feature = "unstable-pre-spec")]
+    #[serde(rename = "m.new_content", skip_serializing_if = "Option::is_none")]
+    pub new_content: Option<Box<MessageEventContent>>,
+}
+
+impl MessageEventContent {
+    /// Create a `MessageEventContent` with the given `MessageType`.
+    pub fn new(msgtype: MessageType) -> Self {
+        Self {
+            msgtype,
+            relates_to: None,
+            #[cfg(feature = "unstable-pre-spec")]
+            new_content: None,
+        }
+    }
+
+    /// A constructor to create a plain text message.
+    pub fn text_plain(body: impl Into<String>) -> Self {
+        Self {
```

Comment:
The `text_` and `notice_` constructors can all be simplified to `Self::new(...)`.